packages/flutter/lib/src/widgets/sliver_coordinator.dart

```diff
@@ -0,0 +1,254 @@
+/// whenever a scroll-end gesture leaves it partially visible. The
+/// auto-scroll is triggered by the [SliverCoordinator]'s
+/// callback. The callback has access to the current scroll offset and
+/// the current extent (height) of the [CoordinatedSliver].
+///
+/// ** See code in examples/api/lib/widgets/sliver/sliver_coordinator.0.dart **
+/// {@end-tool}
+class CoordinatedSliver extends SingleChildRenderObjectWidget {
+  /// Encapsulate a Sliver and save its layout information so that it can be
+  /// retrieved from a [SliverCoordinator.callback] after its [CustomScrollView]
+  /// has completed layout.
+  const CoordinatedSliver({
+    super.key,
+    Widget? sliver,
+  }) : super(child: sliver);
+
+  @override
+  RenderObject createRenderObject(BuildContext context) {
+    return RenderCoordinatedSliver(
+      id: this,
+      data: SliverCoordinator._of(context),
+    );
+  }
+
+  @override
+  void updateRenderObject(BuildContext context, RenderCoordinatedSliver renderObject) {
+    renderObject.id = this;
+    renderObject.data = SliverCoordinator._of(context);
+  }
+
+  @override
+  void didUnmountRenderObject(RenderCoordinatedSliver renderObject) {
+    renderObject.data._remove<_SliverLayoutInfo>(this);
+  }
+
+  /// True if [getSliverConstraints] and [getSliverGeometry] are safe to call
+  /// from  [SliverCoordinator.callback].
+  ///
+  /// False if this sliver wasn't laid out in the most recent [CustomScrollView]
+  /// layout.
+  bool hasLayoutInfo(SliverCoordinatorData data) {
+    return data._get<_SliverLayoutInfo>(this) != null;
+  }
+
+  /// The [SliverConstraints] computed for this sliver in the most recent
+  /// [CustomScrollView] layout.
+  SliverConstraints getSliverConstraints(SliverCoordinatorData data) {
+    assert(hasLayoutInfo(data));
+    return data._get<_SliverLayoutInfo>(this)!.constraints;
+  }
+
+  /// The [SliverGeometry] computed for this sliver in the most recent
+  /// [CustomScrollView] layout.
+  SliverGeometry getSliverGeometry(SliverCoordinatorData data) {
+    assert(hasLayoutInfo(data));
+    return data._get<_SliverLayoutInfo>(this)!.geometry;
+  }
+}
+
+
+/// A sliver that saves its [SliverConstraints] and [SliverGeometry]
+/// layout information in [data], using [id] as a key, each time it's laid out.
+class RenderCoordinatedSliver extends RenderProxySliver {
+  /// Creates a render object that that saves its [SliverConstraints] and [SliverGeometry]
+  /// values in a [SliverCoordinatorData] table each time it's laid out.
+  RenderCoordinatedSliver({
+    required this.id,
+    required this.data,
+    RenderSliver? child,
+  }) : super(child);
+
+  /// The [SliverCoordinatorData] table lookup key for this sliver's layout
+  /// information.
+  ///
+  /// Currently the [CoordinatedSliver] itself is used as its lookup key.
+  Object id;
+
+  /// The [SliverCoordinatorData] table where this sliver's layout information
+  /// will be stored.
+  SliverCoordinatorData data;
```

Comment:
Should these have getters and setters that trigger layout if they change? Or maybe in the case of the id changing, would we need a _replace or _update method on SliverCoordinatorData?

Reply:
Dynamic updates to this table aren't needed. The SliverCoordinatorData's internal table which maps from ids to sliver geometry and constraints, is cleared when layout begins. It's used to lookup geometry and constraints for the most recent layout and whatever ids were in use at that time.

Reply:
> is cleared when layout begins.

Right, but how would layout be triggered? Same with id, when updateRenderObject is called, if data or id change, this will not trigger a layout for the coordinated sliver.

Reply:
Changing a CoordinatedSliver's id or its data should not trigger a layout.  When a layout does happen (for the usual reasons) the updated id and data will be visible to the CoordinatedSliver's callback.

Reply:
What if layout ides are swapped? It won't trigger the callback until scrolling occurs.